### PR TITLE
feat(dolt): surface active compaction quarantine markers in gc dolt health

### PR DESCRIPTION
**Pull request by rjgeng**
## Summary

An active compaction quarantine marker
(`$PACK_STATE_DIR/compact-quarantine/<db>`, written by `gc dolt compact` when a
post-flatten integrity probe trips) blocks auto-GC and scheduled compaction for
that database **indefinitely**, but nothing in `gc dolt health` surfaced it. A
marker could sit for many days while the un-GC'd working set grew unbounded and
degraded the managed `dolt sql-server` — with no signal pointing at the
quarantine; it only became visible on a manual `gc dolt compact`.

Addresses #3729 (the primary observability ask).

## What changed

`gc dolt health` now scans `compact-quarantine/` — filesystem-only, independent
of server reachability, since the un-GC'd bloat can itself wedge the server —
and reports each marker's db, reason, and age:

- **human mode:** a `Compaction quarantine: N (auto-GC blocked)` section, one
  line per db (`<db>: <reason> (held <age>)`). Exits a distinct code **2**
  (reachable but quarantined) so CLI/CI callers catch a blocked compaction
  without conflating it with an unreachable server (exit 1).
- **`--json` mode:** a `quarantine` array of `{db, reason, age_sec}`. Stays
  **exit 0** so programmatic consumers (`gc dolt health-check`, which keys only
  on `server.reachable`) parse the payload unchanged — non-breaking.

The marker directory, one-file-per-db layout, and `key=value` body are read
exactly as `commands/compact/run.sh` writes them.

## Scope

Single-purpose: surfacing the quarantine. The secondary `Backups: none found`
false-negative mentioned in #3729 (health gauges backups by globbing
`migration-backup-*`, ignoring periodic `.dolt-backup/<db>` artifacts) is left
for a follow-up — different code path, easier to review separately.

## Exit-code note for reviewers

The new exit **2** is the one behavior addition beyond pure reporting. It is
opt-in for the human/CLI path only; `--json` is untouched (still 0), so the
`dolt-health` order and `gc dolt health-check` are unaffected. Happy to drop it
to warn-only if you'd prefer health stay exit-0 on a quarantine.

## Tests

- `TestHealthScriptSurfacesQuarantineInJSON` — the `quarantine` array carries
  db/reason and a `created_at`-derived `age_sec`.
- `TestHealthScriptQuarantineHumanExitCode` — exit 2 + section when a marker is
  present; exit 0 + silent when absent.

### CODE REVIEW

File: examples/bd/dolt/commands/health/run.sh [hunk 1]
```diff
@@ -92,6 +93,33 @@ now_ms() {
...
+marker_epoch() {
+  _ts="$1"
+  case "$_ts" in
+    ''|*[!0-9TZ:.+-]*) return ;;
+  esac
+  # GNU date parses the RFC3339 string directly; BSD/macOS date needs an
+  # explicit input format and the -j (do-not-set-clock) flag.
+  _e=$(date -u -d "$_ts" +%s 2>/dev/null) && { printf '%s' "$_e"; return; }
+  _e=$(date -u -j -f "%Y-%m-%dT%H:%M:%SZ" "$_ts" +%s 2>/dev/null) && { printf '%s' "$_e"; return; }
+}
```
[the remark this answers is not in the document]
Comment (rjgeng) [reply examples/bd/dolt/commands/health/run.sh:111]: Good catch, thank you. You're right: both `&&` chains set a non-zero exit status when `date` fails to parse the timestamp, and since the function is called via command substitution in a script running `set -e`, that non-zero propagates and aborts the script before the `stat`-mtime fallback fires.

Fixed by replacing the `&&` forms with `if` conditionals and adding an explicit `return 0` at the end of the function — `if cmd; then ...; fi` exempts `cmd` from `set -e` in all POSIX-compliant shells, so the function now exits 0 regardless of which `date` variant fails. Full test suite still passes.

**Comment by wbern** [comment]
This would resolve #3729. Flagging that the demand signal there has grown — there's additional field data on the issue describing a multi-day silent compaction-quarantine stall that `gc doctor`'s current checks didn't surface. Happy to help test against a repro if that's useful.

**Comment by rjgeng** [comment]
Thanks for flagging this, @wbern — good to know the demand signal on #3729 is real and growing (matches the field report `tarunmistry` posted there too).

Quick status: the one substantive issue raised in review here (Copilot: `marker_epoch()` could return non-zero on a corrupted/unparseable timestamp, which would abort the whole health check under `set -e` before it reached the file-mtime fallback) was fixed early on and is confirmed still correct at the current head — the function now returns 0 unconditionally regardless of which `date` variant fails.

Scope is deliberately narrow to surfacing the quarantine state itself; the `Backups: none found` false-negative some of you may be hitting alongside it (health only globs `migration-backup-*`, not periodic `.dolt-backup/<db>` artifacts) is intentionally left for a separate follow-up.

If you're able to test against a real quarantine repro, that'd genuinely help move this forward — particularly any edge case the exit-code split (2 = reachable-but-quarantined, 1 = unreachable) doesn't handle cleanly for your setup.

**Review by quad341 — APPROVED** [review]
@rjgeng — this is exactly the kind of PR I like reviewing. Every load-bearing claim in your description holds up under tracing: the quarantine gate in `compact/run.sh` really is a bare marker check with no age expiry, so "blocked indefinitely" is precise, and the `dolt-health` order really does go `--json | health-check`, which keys only on `server.reachable` — so your exit-2 split genuinely can't destabilize the patrol. Scanning filesystem-only because the un-GC'd bloat can wedge the server is the right instinct.

One thing I'll take on my side, no action needed from you: `compact/run.sh` drops transient `<db>.probe.XXXXXX` files into that same directory on *every* flatten, and the glob reads them as markers — a phantom entry and a false exit 2. It's a one-line filter plus a test, and I'll document `quarantine` in `result.schema.json` while I'm in there.

The exit-2 question I'm routing to a maintainer, since it's a CLI contract — thank you for flagging it yourself and offering the warn-only fallback. That's the good version of asking. The `reachableServerEnv` refactor is a nice bonus too.


---
**Reviewers**: Qwen (local) — ok · Claude (claude-opus-5[1m]) — ok · Codex (gpt-5.6-sol) — ok
**Synthesis**: claude-opus-5[1m]
